cpp/src/arrow/scalar.h

```diff
@@ -46,11 +46,13 @@ class Array;
 struct ARROW_EXPORT Scalar {
   virtual ~Scalar() = default;
 
+  explicit Scalar(const std::shared_ptr<DataType>& type) : type(type), is_valid(false) {}
```

Comment:
This constructor should be protected since an instance of `Scalar` (not one of its subclasses) should never be constructed

Reply:
Good point.